atsign-foundation/at_server
# feat: APKAM sync changes

**- What I did**
- When connection is authenticated via the APKAM
   - Enhance the sync verb to send the keys associated with enrolled namespaces in the sync response for the client to sync.
   - Enhance the "stats" verb to provide the highest commitId among the enrolled namespaces.

**- How I did it**
- Within the `SyncProgressiveVerbHandler`,`prepareResponse` method, when the connection is authenticated through APKAM, retrieve the enrolled namespaces. If the key's namespace is found among the enrolled namespaces or if the enrolled namespaces contain "*", include the key entry in the sync response for client to sync.

- With the above changes, the following scenario causes the sync to run into infinite loop : If a client submits a regex containing two namespaces, where one of the namespaces lacks authorization, and the key with the unauthorized namespace possesses the highest commitId, it can potentially lead to an infinite loop within the client's sync service. This issue arises because the server fails to send the commit entry with the highest commitId. To address the problem, enhance the "stats:3" verb to send the latestCommitId among the enrolled namespaces.

- The stats_verb_handler, call the "getMetrics" method to fetch the latestCommitId. In the `getMetrics` method within the `LastCommitIDMetricImpl`, introduce an optional parameter called `enrolledNamespaces` that is set within the `stats_verb_handler` when the connection is authenticated via APKAM.

- Modify the `getMetrics` method to pass the `enrolledNamespaces` parameter to the `atCommitLog->lastCommittedSequenceNumberWithRegex` method in order to filter the keys based on the enrolled namespaces.

- In the `CommitLogKeyStore.dart` file, include a new method named `isNamespaceAuthorized` to validate whether the key's namespace is authorized.

**- How to verify it**
- Added unit tests in "stats_verb_test.dart" and "sync_verb_test.dart"
- Added functional test in "enroll_verb_test.dart" to verify highest commit id among the enrolled namespaces is returned
- Added functional test in "sync_verb_test.dart" to verify keys with enrolled namespaces are returned in the sync response

**- Description for the changelog**
- Implement APKAM changes into sync verb

NOTE: 
1. Added a work around https://github.com/atsign-foundation/at_server/pull/1569/commits/e9647ae2b6c19171e9746e121493d99e5b66b3fd for the bug: https://github.com/atsign-foundation/at_server/issues/1570 in "commit_log_keystore.dart"
2. The changes in at_persistence_secondary_server which are a part of this PR are also in "apkam_sync_persistence_changes" which has CHANGELOG and pubspec version updated.
